File: backend/authentication-service/src/main/java/com/example/demo/services/UserService.java
package com.example.demo.services;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.example.demo.dto.UserDTO;
import com.example.demo.dto.UserRegisterDTO;
import com.example.demo.entities.Role;
import com.example.demo.entities.User;
import com.example.demo.entities.User.Status;
import com.example.demo.exceptions.EmailAlreadyVerifiedException;
import com.example.demo.exceptions.InvalidEmailException;
import com.example.demo.repositories.RoleRepository;
import com.example.demo.repositories.UserRepository;

@Service
public class UserService {

    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[A-Za-z0-9+_.-]+@(.+)$");

    @Autowired
    private UserRepository urepo;

    @Autowired
    private RoleRepository rrepo;

    @Autowired
    private PasswordEncoder encoder;

    @Autowired
    private VerificationJwtService verificationJwtService;

    @Autowired
    private EmailService emailService;


    /**
     * Registers a new user
     */
    public boolean addUser(UserRegisterDTO userdto) {
        if (userdto == null) {
            throw new IllegalArgumentException("User registration data cannot be null");
        }
        if (userdto.getName() == null || userdto.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Name is required");
        }
        if (userdto.getEmail() == null || userdto.getEmail().trim().isEmpty()) {
            throw new IllegalArgumentException("Email is required");
        }
        if (userdto.getPassword() == null || userdto.getPassword().trim().isEmpty()) {
            throw new IllegalArgumentException("Password is required");
        }
        if (userdto.getPhone() == null || userdto.getPhone().trim().isEmpty()) {
            throw new IllegalArgumentException("Phone number is required");
        }

        if (urepo.existsByEmail(userdto.getEmail().trim())) {
            throw new IllegalArgumentException("Email is already registered");
        }

        if (urepo.existsByPhone(userdto.getPhone().trim())) {
            throw new IllegalArgumentException("Phone number is already registered");
        }

        Role role = rrepo.findById(userdto.getRid())
                .orElseThrow(() -> new IllegalArgumentException("Role not found for ID: " + userdto.getRid()));

        User user = new User();
        user.setRole(role);
        user.setName(userdto.getName().trim());
        user.setEmail(userdto.getEmail().trim());
        user.setPhone(userdto.getPhone().trim());
        user.setPassword(encoder.encode(userdto.getPassword()));
        user.setAddress(userdto.getAddress());
        user.setCity(userdto.getCity());
        user.setState(userdto.getState());
        user.setCountry(userdto.getCountry());
        user.setImage(null);
        user.setStatus(Status.Active);

        try {
            User savedUser = urepo.save(user);
            System.out.println("User Saved Successfully: " + savedUser.getEmail());
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException("Failed to save user: " + e.getMessage(), e);
        }
    }

    /**
     * Fetch user by email
     */
    public User getUser(String email) {
        return urepo.findByEmail(email);
    }

    // update user
    public boolean updateUser(Integer uid, UserRegisterDTO userdto) {

        if (uid == null) {
            throw new IllegalArgumentException("User id is required");
        }

        if (userdto == null) {
            throw new IllegalArgumentException("User data cannot be null");
        }

        User user = urepo.findById(uid)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        // Name
        if (userdto.getName() == null || userdto.getName().trim().isEmpty()) {
            throw new IllegalArgumentException("Name is required");
        }
        user.setName(userdto.getName().trim());

        userdto.setEmail(user.getEmail());

        user.setPhone(user.getPhone());

        // Password (Update only if provided)
        if (userdto.getPassword() != null && !userdto.getPassword().trim().isEmpty()) {
            user.setPassword(encoder.encode(userdto.getPassword()));
        }

        // Address Details
        user.setAddress(userdto.getAddress());
        user.setCity(userdto.getCity());
        user.setState(userdto.getState());
        user.setCountry(userdto.getCountry());

        try {
            urepo.save(user);
            return true;
        } catch (Exception e) {
            throw new RuntimeException("Failed to update user: " + e.getMessage(), e);
        }
    }

    public User updateUser(int id, com.example.demo.dto.UserDTO userdto) {
        User user = urepo.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        user.setName(userdto.getName());
        user.setPhone(userdto.getPhone());
        user.setAddress(userdto.getAddress());
        user.setCity(userdto.getCity());
        user.setState(userdto.getState());
        user.setCountry(userdto.getCountry());
        return urepo.save(user);
    }

    public User getUserById(Integer uid) {
        return urepo.findById(uid)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public boolean changePassword(int id, String oldPassword, String newPassword) {
        User user = urepo.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        if (!encoder.matches(oldPassword, user.getPassword())) {
            return false;
        }
        user.setPassword(encoder.encode(newPassword));
        urepo.save(user);
        return true;
    }

    public Map<String, Object> getUserCounts() {
        long totalUsers = urepo.count();
        long candidateCount = urepo.countByRoleName("Candidate");
        long employerCount = urepo.countByRoleName("Employer");
        long adminCount = urepo.countByRoleName("Admin");

        Map<String, Object> counts = new HashMap<>();
        counts.put("totalUsers", totalUsers);
        counts.put("candidateCount", candidateCount);
        counts.put("employerCount", employerCount);
        counts.put("adminCount", adminCount);
        return counts;
    }

    public List<UserDTO> getAllUsers() {
        return urepo.findAll().stream().map(user -> new UserDTO(
                user.getUid(),
                user.getName(),
                user.getEmail(),
                user.getPhone(),
                user.getAddress(),
                user.getCity(),
                user.getState(),
                user.getCountry(),
                user.getRole() != null ? user.getRole().getRname() : "N/A")).collect(Collectors.toList());
    }

    /**
     * Sends a 6-digit verification OTP to the user's email and returns a Verification JWT.
     */
    public String sendVerificationOtp(String email) {
        if (email == null || email.trim().isEmpty() || !EMAIL_PATTERN.matcher(email.trim()).matches()) {
            throw new InvalidEmailException("Invalid email format or missing email address.");
        }

        String trimmedEmail = email.trim();

        // Check if user already exists and is already verified
        User existingUser = urepo.findByEmail(trimmedEmail);
        if (existingUser != null && existingUser.isEmailVerified()) {
            throw new EmailAlreadyVerifiedException("This email address is already verified.");
        }

        String plainOtp = verificationJwtService.generate6DigitOtp();
        String verificationJwt = verificationJwtService.createVerificationToken(trimmedEmail, plainOtp);

        emailService.sendOtpEmail(trimmedEmail, plainOtp);

        return verificationJwt;
    }

    /**
     * Verifies the OTP and token, marking the user's email as verified in the DB if present.
     */
    public boolean verifyEmailOtp(String email, String otp, String token) {
        if (email == null || email.trim().isEmpty() || !EMAIL_PATTERN.matcher(email.trim()).matches()) {
            throw new InvalidEmailException("Invalid email format or missing email address.");
        }

        String trimmedEmail = email.trim();

        // Validate JWT and OTP hash match
        verificationJwtService.validateAndVerifyOtp(trimmedEmail, otp, token);

        // Update database if user exists
        User user = urepo.findByEmail(trimmedEmail);
        if (user != null) {
            if (user.isEmailVerified()) {
                throw new EmailAlreadyVerifiedException("This email address is already verified.");
            }
            user.setEmailVerified(true);
            urepo.save(user);
        }

        return true;
    }
}

File: backend-dotnet/Services/JobPortal.AuthService/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using JobPortal.Shared.Data;
using JobPortal.Shared.DTOs;
using JobPortal.Shared.Entities;

namespace JobPortal.AuthService.Services;

public class UserService
{
    private readonly JobPortalDbContext _context;

    public UserService(JobPortalDbContext context)
    {
        _context = context;
    }

    public async Task<bool> AddUserAsync(UserRegisterDTO dto)
    {
        if (dto == null) throw new ArgumentNullException(nameof(dto));
        if (string.IsNullOrWhiteSpace(dto.Name)) throw new ArgumentException("Name is required");
        if (string.IsNullOrWhiteSpace(dto.Email)) throw new ArgumentException("Email is required");
        if (string.IsNullOrWhiteSpace(dto.Password)) throw new ArgumentException("Password is required");
        if (string.IsNullOrWhiteSpace(dto.Phone)) throw new ArgumentException("Phone number is required");

        var cleanEmail = dto.Email.Trim();
        var cleanPhone = dto.Phone.Trim();

        if (await _context.Users.AnyAsync(u => u.Email == cleanEmail))
        {
            throw new ArgumentException("Email is already registered");
        }

        if (await _context.Users.AnyAsync(u => u.Phone == cleanPhone))
        {
            throw new ArgumentException("Phone number is already registered");
        }

        var role = await _context.Roles.FindAsync(dto.Rid)
            ?? throw new ArgumentException($"Role not found for ID: {dto.Rid}");

        var user = new User
        {
            Name = dto.Name.Trim(),
            Email = cleanEmail,
            Phone = cleanPhone,
            Password = BCrypt.Net.BCrypt.HashPassword(dto.Password),
            Address = dto.Address,
            City = dto.City,
            State = dto.State,
            Country = dto.Country,
            Rid = dto.Rid,
            Status = "Active",
            CreatedAt = DateTime.UtcNow
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<User?> GetUserByEmailAsync(string email)
    {
        return await _context.Users
            .Include(u => u.Role)
            .FirstOrDefaultAsync(u => u.Email == email);
    }

    public async Task<User?> GetUserByIdAsync(int uid)
    {
        return await _context.Users
            .Include(u => u.Role)
            .FirstOrDefaultAsync(u => u.Uid == uid);
    }

    public async Task<bool> UpdateUserAsync(int uid, UserRegisterDTO dto)
    {
        var user = await _context.Users.FindAsync(uid)
            ?? throw new ArgumentException("User not found");

        if (string.IsNullOrWhiteSpace(dto.Name)) throw new ArgumentException("Name is required");

        user.Name = dto.Name.Trim();

        if (!string.IsNullOrWhiteSpace(dto.Password))
        {
            user.Password = BCrypt.Net.BCrypt.HashPassword(dto.Password);
        }

        user.Address = dto.Address;
        user.City = dto.City;
        user.State = dto.State;
        user.Country = dto.Country;

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<User> UpdateUserDtoAsync(int id, UserDTO dto)
    {
        var user = await _context.Users.Include(u => u.Role).FirstOrDefaultAsync(u => u.Uid == id)
            ?? throw new KeyNotFoundException("User not found");

        user.Name = dto.Name;
        user.Phone = dto.Phone;
        user.Address = dto.Address;
        user.City = dto.City;
        user.State = dto.State;
        user.Country = dto.Country;

        await _context.SaveChangesAsync();
        return user;
    }

    public async Task<bool> ChangePasswordAsync(int id, string oldPassword, string newPassword)
    {
        var user = await _context.Users.FindAsync(id)
            ?? throw new KeyNotFoundException("User not found");

        if (!BCrypt.Net.BCrypt.Verify(oldPassword, user.Password))
        {
            return false;
        }

        user.Password = BCrypt.Net.BCrypt.HashPassword(newPassword);
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<Dictionary<string, object>> GetUserCountsAsync()
    {
        var totalUsers = await _context.Users.CountAsync();
        var candidateCount = await _context.Users.CountAsync(u => u.Role != null && (u.Role.Rname == "Candidate" || u.Role.Rname == "CANDIDATE"));
        var employerCount = await _context.Users.CountAsync(u => u.Role != null && (u.Role.Rname == "Employer" || u.Role.Rname == "EMPLOYER" || u.Role.Rname == "Recruiter" || u.Role.Rname == "RECRUITER"));
        var adminCount = await _context.Users.CountAsync(u => u.Role != null && (u.Role.Rname == "Admin" || u.Role.Rname == "ADMIN"));

        return new Dictionary<string, object>
        {
            { "totalUsers", totalUsers },
            { "candidateCount", candidateCount },
            { "employerCount", employerCount },
            { "adminCount", adminCount }
        };
    }

    public async Task<List<UserDTO>> GetAllUsersAsync()
    {
        var users = await _context.Users.Include(u => u.Role).ToListAsync();
        return users.Select(u => new UserDTO(
            u.Uid,
            u.Name,
            u.Email,
            u.Phone,
            u.Address,
            u.City,
            u.State,
            u.Country,
            u.Role?.Rname ?? "N/A"
        )).ToList();
    }
}
